test: deduplicate reorg test code

yuvicc opened the pull request:
Follow-up to #32587. 

`trigger_reorg()` was copy-pasted across several functional tests. This moves it into `blocktools.py`, together with the fork construction, as two helpers: `create_empty_fork()` and `trigger_reorg()`. 

The remaining tests that still simulated reorgs with the `invalidateblock` RPC are converted to use them, since `invalidateblock` takes a different codepath than a reorg caused by a competing chain arriving.

maflcko commented:
are you still working on this, or can it be closed?

yuvicc commented:
> are you still working on this, or can it be closed?

Yes resumed the work, will push the update soon!

yuvicc commented:
Updated the PR description and ready for review

sedited commented:
@instagibbs do you want to take a look here?

instagibbs commented:
I think this might be a bit of overengineering. Centralizing code is nice, but really I don't see the value of the added abstraction aside from centralizing trigger_reorg?

yuvicc committed:
test: move duplicated `trigger_reorg` method to blocktools.py.

yuvicc committed:
test: fix reorg behaviour in wallet, p2p and rpc functional tests to use proper fork-based approach

yuvicc commented:
Addressed @instagibbs suggestion to only keep `trigger_reorg` method in blocktools.py.

instagibbs commented on test/functional/p2p_sendheaders.py:
```diff
@@ -208,17 +208,17 @@ def mine_reorg(self, length):
+        # Prep for fork with empty blocks
+        fork_blocks = create_empty_fork(self.nodes[1], fork_length=length + 1)
+
         # make sure all invalidated blocks are node0's
```
stale comment with "invalid"

instagibbs commented on test/functional/wallet_balance.py:
```diff
@@ -237,13 +241,17 @@ def test_balances(*, fee_node_1=0):
         self.nodes[1].sendrawtransaction(hexstring=tx_replace, maxfeerate=0)
 
+        # Prep for fork
+        fork_blocks0 = create_empty_fork(self.nodes[0])
```
are these nodes disconnected? I think it can be prepped once and used to trigger reorg on both sides?

yuvicc replied:
Correct, they are connected. Will do the changes.

instagibbs commented on test/functional/rpc_gettxspendingprevout.py:
```diff
@@ -184,7 +189,10 @@ def create_tx(**kwargs):
         assert_equal(result, [spent_out_in_block(tx1['txid'], vout=0, spending_tx_id=tx2["txid"], blockhash=blockhash, spending_tx=tx2['hex'])])
 
-        # replace tx1 with tx3
+        # replace tx1 with tx3. We use invalidateblock (rather than a fork-based
```
suggestion for wording:

          # Disconnect tx1's block so tx1 can be replaced by tx3 below, which also
          # evicts tx1's child tx2. Unlike the case above, use invalidateblock rather
          # than a fork-based reorg: the index is only rewound when a block connects
          # so disconnecting without connecting anything lets us check below that tx2 is still indexed until the next block.

instagibbs commented on test/functional/wallet_balance.py:
```diff
@@ -237,13 +241,17 @@ def test_balances(*, fee_node_1=0):
+
         # Now confirm tx_replace
-        block_reorg = self.generatetoaddress(self.nodes[1], 1, ADDRESS_WATCHONLY)[0]
+        self.generatetoaddress(self.nodes[1], 1, ADDRESS_WATCHONLY)[0]
```
no need for [0] subscript anymore

instagibbs commented on test/functional/mempool_updatefromblock.py:
```diff
@@ -47,8 +44,7 @@ def transaction_graph_test(self, size, *, n_tx_to_mine, fee=100_000):
         wallet = MiniWallet(self.nodes[0])
 
-        # Prep for fork with empty blocks to not use invalidateblock directly
-        # for reorg case. The rpc has different codepath
```
wondering if there's a good place to add this removed comment to have people not reintroduce this pattern

yuvicc replied:
Maybe we can add in `create_empty_fork` docstring?

<details>

```python
def create_empty_fork(node, fork_length=FORK_LENGTH):
    '''
        Creates a fork using node's chaintip as the starting point.
        Returns a list of blocks to submit in order.

        Prefer this over calling the invalidateblock RPC to simulate a reorg:
        invalidateblock takes a different codepath than a reorg caused by a
        competing chain arriving, so tests using it are not exercising the
        behaviour they mean to.
    '''
```

</details>

yuvicc commented:
Thanks for the review @instagibbs 
I've removed `Forkgenerator` class and only kept `trigger_reorg` in blocktools and addressed some comments.

I have also updated PR title and the description.
